microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/ConfigSourceRootRegistrationService.java

```diff
@@ -0,0 +1,119 @@
+ */
+class ConfigSourceRootRegistrationService implements Service {
+
+    private final String name;
+    private final String path;
+    /**
+     * Can be null
+     */
+    private final String relativeTo;
+    private final int ordinal;
+    private final Supplier<PathManager> pathManager;
+    private final Registry<ConfigSourceProvider> providerRegistry;
+
+    ConfigSourceRootRegistrationService(String name, String path, String relativeTo, int ordinal, Supplier<PathManager> pathManager, Registry<ConfigSourceProvider> sources) {
+        this.name = name;
+        this.path = path;
+        this.relativeTo = relativeTo;
+        this.ordinal = ordinal;
+        this.pathManager = pathManager;
+        this.providerRegistry = sources;
+    }
+
+    static void install(OperationContext context, String name, String path, String relativeTo, int ordinal, Registry<ConfigSourceProvider> registry) {
+        ServiceBuilder<?> builder = context.getServiceTarget()
+                .addService(ServiceNames.CONFIG_SOURCE_ROOT.append(name));
+        Supplier<PathManager> pathManager = builder.requires(context.getCapabilityServiceName("org.wildfly.management.path-manager", PathManager.class));
+
+        builder.setInstance(new ConfigSourceRootRegistrationService(name, path, relativeTo, ordinal, pathManager, registry))
+                .install();
+    }
+
+    @Override
+    public void start(StartContext startContext) {
+        String relativeToPath = AbsolutePathService.isAbsoluteUnixOrWindowsPath(path) ? null  : relativeTo;
+        String dirPath = pathManager.get().resolveRelativePathEntry(path, relativeToPath);
+        File dir = new File(dirPath);
+        MicroProfileConfigLogger.ROOT_LOGGER.loadConfigSourceRootFromDir(dir.getAbsolutePath());
+        this.providerRegistry.register(this.name, new RootDirectoryConfigSourceProvider(name, dir, ordinal));
+    }
+
+    @Override
+    public void stop(StopContext context) {
+        this.providerRegistry.unregister(this.name);
+    }
+
+    private static class RootDirectoryConfigSourceProvider implements ConfigSourceProvider {
+        private final String name;
+        private final File rootDirectory;
+        private int ordinal;
+
+        public RootDirectoryConfigSourceProvider(String name, File rootDirectory, int ordinal) {
+            this.name = name;
+            this.rootDirectory = rootDirectory;
+            this.ordinal = ordinal;
+        }
+
+        @Override
+        public Iterable<ConfigSource> getConfigSources(ClassLoader forClassLoader) {
+            TreeMap<String, ConfigSource> map = new TreeMap<>();
+            File[] files = rootDirectory.listFiles();
+            if (files != null) {
+                // If the directory exist, this will be an empty list
+                for (File file : rootDirectory.listFiles()) {
+                    if (file.isDirectory()) {
+                        map.put(file.getName(), new FileSystemConfigSource(file, ordinal));
```

Comment:
Should we log all the top-level dir that would be used as config sources.

I think that'd be valuable info to see something in the log like:

```
* INFO Using /etc/config/rhosak (ordinal=100) as ConfigSource
* INFO Using /etc/config/mydb-secret (ordinal=102) as ConfigSource
```

Reply:
Done